File: agent/src/services/docker.service.ts
import Docker from "dockerode";
import { ContainerInfo } from "dockerode";
import { IDockerContainer } from "../interfaces/docker-container.interface";
import { formatDockerUptime } from "../helpers/format-docker-uptime";

const docker = new Docker({
  socketPath: "/var/run/docker.sock",
});

export class DockerService {
  static async getContainers(): Promise<IDockerContainer[]> {
    const containers: ContainerInfo[] = await docker.listContainers({
      all: true,
    });

    return containers.map((container: ContainerInfo) => ({
      id: container.Id,
      name: container.Names?.[0]?.replace("/", "") || "unknown",
      image: container.Image,
      state: container.State,
      status: formatDockerUptime(container.Status),
    }));
  }
}

File: agent-go/services/docker_service.go
package services

import (
	"context"

	"agent-go/helpers"
	"agent-go/types"

	containerTypes "github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"
)

func GetContainers() ([]types.DockerContainer, error) {
	ctx := context.Background()

	cli, err := client.NewClientWithOpts(
		client.FromEnv,
		client.WithAPIVersionNegotiation(),
	)

	if err != nil {
		return nil, err
	}

	containers, err := cli.ContainerList(
		ctx,
		containerTypes.ListOptions{
			All: true,
		},
	)

	if err != nil {
		return nil, err
	}

	result := []types.DockerContainer{}

	for _, container := range containers {
		name := "unknown"

		if len(container.Names) > 0 {
			name = container.Names[0][1:]
		}

		result = append(result, types.DockerContainer{
			ID:     container.ID,
			Name:   name,
			Image:  container.Image,
			State:  container.State,
			Status: helpers.FormatDockerUptime(container.Status),
		})
	}

	return result, nil
}